# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 讀取數據

In [1]:
import pandas as pd

In [2]:
# from google.colab import files  ###要把檔案上傳到colab才可以讀取文件
# uploaded = files.upload()

In [3]:
original_data = pd.read_csv(r"C:\Users\User\Desktop\python\e_commerce.csv")

In [4]:
original_data = pd.read_csv("e_commerce.csv")
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 評估數據

###評估數據整齊度

In [5]:
original_data.sample(10)

InvoiceNo StockCode                         Description  Quantity  \
137208    548148     22721       SET OF 3 CAKE TINS SKETCHBOOK         1   
492110    578088     21744     SNOWFLAKE PORTABLE TABLE LIGHT          3   
510619    579459     21108  FAIRY CAKE FLANNEL ASSORTED COLOUR        27   
134935    547872     22631            CIRCUS PARADE LUNCH BOX          1   
419892    572862     23356               LOVE HOT WATER BOTTLE         1   
43272     540026    84849B              FAIRY SOAP SOAP HOLDER         1   
52700     540813     22470               HEART OF WICKER LARGE         2   
9989      537224     22766                 PHOTO FRAME CORNICE         4   
530750    580880     21579     LOLITA  DESIGN  COTTON TOTE BAG        12   
82159     543191     21124  SET/10 BLUE POLKADOT PARTY CANDLES         3   

             InvoiceDate  UnitPrice  CustomerID         Country  
137208   3/29/2011 12:31       4.95     17931.0  United Kingdom  
492110  11/22/2011 16:47       2.95     16376.0  United Kingdom  
510619  11/29/2011 14:38       2.10     17173.0  United Kingdom  
134935   3/27/2011 14:34       1.95     16777.0  United Kingdom  
419892  10/26/2011 12:48       5.95     16750.0  United Kingdom  
43272     1/4/2011 13:25       3.36         NaN  United Kingdom  
52700    1/11/2011 12:41       2.95     14669.0  United Kingdom  
9989     12/5/2010 16:24       2.95     13174.0  United Kingdom  
530750   12/6/2011 12:19       2.25     14121.0  United Kingdom  
82159     2/4/2011 11:54       1.25     13137.0  United Kingdom

抽樣10行數據符合"每列一個變量，每行一個觀察值，每個單元格是一個值"，具體來看每行是關於某商品的一次交易，每列是交易相關的各個變量，因此不存在結構性問題。

###評估數據乾淨度

In [6]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


數據總共有541909行`("541909 entries")`

`Description`,`CustomerID`有缺失值(不等於總行數)

`InvoiceDate`數據類型應為日期，`CustomerID`數據類型為字符串

###評估缺失數據

In [7]:
original_data[original_data["Description"].isnull()] ### 了解Description存在缺失值後，根據條件提取出缺失觀察值；有1454行是空值

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

`UnitPrice`為0，為了驗證是否存在`Description`變量缺失且`UnitPrice`不為0的數據

In [8]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"]!= 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

結果發現`Description`為空值，`UnitPrice`都為0；此外，分析目標要找出暢銷產品，因此缺少關鍵數據的部分可以刪除，以便後續的數據分析

In [9]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

`CustomerID`不是必要分析的重要數據，並且`CustomerID`缺失的其他交易數據仍然有分析價值，所以保留該數據。

###評估重覆數據

根據資料變數的含義來看，雖然InvoiceNo、StockCode和CustomerID都是唯一識別碼，但一次交易可能包含多件商品，因此InvoiceNo可以存在重複，不同交易可以包含同件商品，因此StockCode可以存在重複，顧客可以進行多次交易或下單多個商品，因此CustomerID也可以存在重複。
那麼針對此資料集，我們無需評估重復資料。


###評估不一致

In [10]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


從`Country`變數值來看，`"USA"`、`"United States"`均在表示美國，`"United Kingdom"`、`"UK"`、`"U.K."`均在表示英國，因此應該在清洗步驟對這些值進行統一，只保留一個指代值。

###評估無效或錯誤數據

可以通過DataFrame的describe方法，對數值統計資訊進行快速瞭解。

In [11]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

從輸出結果來看，`Quantity`和`UnitPrice`存在負數，會對後續數值分析產生影響。

因此，我們先篩選出`Quantity`數值為負數的觀察值，進一步評估其含義。


In [12]:
original_data[original_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

從篩選結果來看，Quantity變數為負數的觀察值，InvoiceNo似乎均以C開頭，表示訂單被取消。
但此猜測不一定準確，為了驗證猜想，我們增加篩選條件，看是否存在Quantity變數為負數且InvoiceNo不以C開頭的觀察值。


In [13]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

"str[0] != "C" "指我們要找出"InvoiceNo"這一列沒有C開頭的數據

以上猜想被證實錯誤，因為還存在Quantity變數為負數且InvoiceNo不以C開頭的觀察值。
但根據以上輸出結果，這些篩選出的觀察值的UnitPrice觀察值均為0，因此增加UnitPrice的條件進行驗證。


In [14]:
original_data[(original_data["Quantity"] < 0)&(original_data["InvoiceNo"].str[0] != "C")&(original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根據輸出結果，猜想得到驗證，當Quantity變數為負數的時候，觀察值滿足以下條件之一：
•	InvoiceNo以C開頭，表示訂單被取消
•	UnitPrice為0，說明單價為0英鎊
這些交易資料均不是有效成交資料，不貢獻銷售，不在後續分析範圍內，因此我們將在資料清理步驟，將Quantity為負數的觀察值刪除。


接下來篩選出UnitPrice為負數的觀察值。

In [15]:
original_data[original_data["UnitPrice"] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

從輸出結果來看，UnitPrice為負數的觀察值都是壞賬調賬(Adjust bad debt)，不屬於實際商品交易資料，因此也在資料清理步驟中也將其刪除。

## 清理数据

根據前面評估部分得到的結論，我們需要進行的資料清理包括：

1.把InvoiceDate變數的資料類型轉換為為日期時間

2.把CustomerID變數的資料類型轉換為字串

3.把Description變數缺失的觀察值刪除

4.把Country變數值"USA"替換為"United States"

5.把Country變數值"UK"、"U.K."替換為"United Kingdom"

6.把Quantity變數值為負數的觀察值刪除

7.把UnitPrice變數值為負數的觀察值刪除

為了區分開經過清理的資料和原始的資料，我們創建新的變數`cleaned_data`，讓它為`original_data`複製出的副本。我們之後的清理步驟都將被運用在`cleaned_data`上。

In [16]:
cleaned_data = original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

1.把InvoiceDate變數的資料類型轉換為為日期時間

In [17]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

2.把CustomerID變數的資料類型轉換為字串

In [18]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].astype(str)
cleaned_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把CustomerID變數值結尾的".0"刪除：

In [19]:
cleaned_data["CustomerID"] = cleaned_data["CustomerID"].str.slice(0, -2)
cleaned_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

.slice(0, -2)：對每個字串做「從 index 0（開頭）到倒數第 2 個字元之前」的切片。

3.把Description變數缺失的觀察值刪除

並查看刪除後該列空缺值個數和：

In [20]:
cleaned_data.dropna(subset=["Description"], inplace=True)

dropna()
是 pandas 用來刪除缺失值（NaN）的函數。

subset=["Description"]
表示只針對 Description 欄位檢查是否為 NaN。

如果 Description 是 NaN → 那一整列（row）會被刪掉。

如果其他欄位是 NaN，但 Description 不是 → 那一列會保留。

In [21]:
cleaned_data["Description"].isnull().sum()

np.int64(0)

把Country變數值"USA"替換為"United States"，並檢查替換後"USA"變數值個數：

In [22]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"USA": "United States"})

In [23]:
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

把Country變數值"UK"、"U.K."替換為"United Kingdom"，並檢查替換後"UK"和"U.K."變數值個數：

In [24]:
cleaned_data["Country"] = cleaned_data["Country"].replace({"UK": "United Kingdom", "U.K.": "United Kingdom"})

In [25]:
print(len(cleaned_data[cleaned_data["Country"] == "UK"]))
print(len(cleaned_data[cleaned_data["Country"] == "U.K."]))

0
0


把Quantity變數值為負數的觀察值刪除，並檢查替換後Quantity變數值為負數的個數：

In [26]:
cleaned_data = cleaned_data[cleaned_data["Quantity"] >= 0]

In [27]:
len(cleaned_data[cleaned_data["Quantity"] < 0])

0

把UnitPrice變數值為負數的觀察值刪除，並檢查替換後UnitPrice變數值為負數的個數：

In [28]:
cleaned_data = cleaned_data[cleaned_data["UnitPrice"] >= 0]

In [29]:
len(cleaned_data[cleaned_data["UnitPrice"] < 0])

0

## 保存清理后的数据

完成資料清理後，把乾淨整齊的資料保存到新的檔裡，檔案名為e_commerce_cleaned.csv。

In [30]:
cleaned_data.to_csv(r"C:\Users\User\Desktop\python\e_commerce_cleaned.csv", index=False)

In [35]:
pd.read_csv(r"C:\Users\User\Desktop\python\e_commerce_cleaned.csv").head()

C:\Users\User\AppData\Local\Temp\ipykernel_4368\402640467.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv(r"C:\Users\User\Desktop\python\e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom